models/migrations/v114.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"net/url"
+	"strings"
+
+	"xorm.io/xorm"
+)
+
+func sanitizeOriginalURL(x *xorm.Engine) error {
+
+	type Repository struct {
+		ID          int64
+		OriginalURL string `xorm:"VARCHAR(2048)"`
+	}
+
+	sess := x.NewSession()
+	defer sess.Close()
+	var last int
+	const batchSize = 50
+	for {
+		var results = make([]Repository, 0, batchSize)
+		err := x.Where("original_url <> '' AND original_url IS NOT NULL").
+			And("original_service_type = 0 OR original_service_type IS NULL").
+			OrderBy("id").
+			Limit(batchSize, last).
+			Find(&results)
+		if err != nil {
+			return err
+		}
+		if len(results) == 0 {
+			break
+		}
+		last += len(results)
+
+		for _, res := range results {
+			u, err := url.Parse(res.OriginalURL)
+			if err != nil {
+				// it is ok to continue here, we only care about fixing URLs that we can read
+				continue
+			}
+
+			if len(u.User.Username()) > 0 {
```

Comment:
u.User is nil?

Reply:
This function already has built in code to check return empty string if u.User is nil:

https://golang.org/src/net/url/url.go?s=11376:11412#L387

So makes one less check for our code needed.

Reply:
OKay, but in fact, I think you could just let `u.User = nil`. See https://goplay.space/#L2yw4PpzBHC

Reply:
ah OK I see yes this is better, added